dom.go

```diff
@@ -709,29 +712,110 @@ func Text(text string, m ...MarkupOrChild) *HTML {
 }
 
 // Rerender causes the body of the given component (i.e. the HTML returned by
-// the Component's Render method) to be re-rendered.
+// the component's Render method) to be re-rendered.
 //
 // If the component has not been rendered before, Rerender panics. If the
 // component was previously unmounted, Rerender is no-op.
+//
+// Rerender operates efficiently by batching renders together. As a result,
+// there is no guarantee that a calls to Rerender will map 1:1 with calls to
+// the component's Render method. For example, two calls to Rerender may
+// result in only one call to the component's Render method.
 func Rerender(c Component) {
 	if c == nil {
 		panic("vecty: Rerender illegally called with a nil Component argument")
 	}
-	prevRender := c.Context().prevRender
-	if prevRender == nil {
-		panic("vecty: Rerender invoked on Component that has never been rendered")
-	}
 	if c.Context().unmounted {
 		return
 	}
-	nextRender, skip, pendingMounts := renderComponent(c, prevRender)
-	if skip {
+	batch.add(c)
+}
+
+// batchRenderer handles component re-renders by queueing and deduplicating
+// them, to be rendered on the next animation frame (via requestAnimationFrame).
+type batchRenderer struct {
+	// batch contains the list of pending components to render.
+	batch []Component
+	// idx maps components to batch indexes to allow dedup, retaining order.
+	idx map[Component]int
+	// scheduled tracks whether a batch has been scheduled for processing.
+	scheduled bool
+}
+
+// add a Component to the pending batch.
+func (b *batchRenderer) add(c Component) {
+	if i, ok := b.idx[c]; ok {
+		// Delete previously queued render.
+		copy(b.batch[i:], b.batch[i+1:])
+		b.batch[len(b.batch)-1] = nil
+		b.batch = b.batch[:len(b.batch)-1]
```

Comment:
I don't think that this code will work as-is.

My read of this code is that you're cutting out the `i`th element. But once you've done this, every other `b.idx` value that comes after `i` becomes invalid. That is, I think you need to update each index in `b.idx` greater than `i` to be `-1` (I could be off-by-one here, so please check my logic).

Reply:
You're right, `idx` needs an update here.